test/e2e/logging_test.go

```diff
@@ -0,0 +1,183 @@
+		})}...)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	url := resources.Route.Status.URL.URL()
+	if _, err := pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		t.Logf,
+		url,
+		v1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(test.HelloWorldText))),
+		"HelloWorldServesText",
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https),
+	); err != nil {
+		t.Fatalf("The endpoint %s for Route %s didn't serve the expected text %q: %v", url, names.Route, test.HelloWorldText, err)
+	}
+
+	pod, err := theOnlyPod(clients, resources.Revision.Namespace, resources.Revision.Name)
+	if err != nil {
+		t.Fatalf("Fail to fetch the pod: %v", err)
+	}
+
+	// A request was sent to / in WaitForEndpointState.
+	if err := waitForLog(clients, pod.Namespace, pod.Name, "queue-proxy", func(log map[string]interface{}) bool {
+		if v, ok := log["requestUrl"]; !ok {
+			return false
+		} else if v.(string) != "/" {
+			return false
+		}
+
+		if v, ok := log["userAgent"]; !ok {
+			return false
+		} else {
+			return v.(string) != network.QueueProxyUserAgent
+		}
+	}); err != nil {
+		t.Fatalf("Got error waiting for normal request logs: %v", err)
+	}
+
+	// Health check requests are sent to / with a specific userAgent value periodically.
+	if err := waitForLog(clients, pod.Namespace, pod.Name, "queue-proxy", func(log map[string]interface{}) bool {
+		if v, ok := log["requestUrl"]; !ok {
+			return false
+		} else if v.(string) != "/" {
+			return false
+		}
+
+		if v, ok := log["userAgent"]; !ok {
+			return false
+		} else {
+			return v.(string) == network.QueueProxyUserAgent
+		}
+	}); err != nil {
+		t.Fatalf("Got error waiting for health check log: %v", err)
+	}
+}
+
+func theOnlyPod(clients *test.Clients, ns, rev string) (corev1.Pod, error) {
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(ns).List(metav1.ListOptions{LabelSelector: fmt.Sprintf("app=%s", rev)})
+
+	if err != nil {
+		return corev1.Pod{}, err
+	}
+
+	if len(pods.Items) != 1 {
+		return corev1.Pod{}, fmt.Errorf("Expect 1 pod, but got %d for %s:%s", len(pods.Items), system.Namespace(), rev)
+	}
+
+	return pods.Items[0], nil
+}
+
+// waitForLog fetches the logs from a container of a pod decided by the given parameters
+// until the given condition is meet or timeout. Most of knative logs are in json format.
+func waitForLog(clients *test.Clients, ns, podName, container string, condition func(log map[string]interface{}) bool) error {
+	return wait.PollImmediate(time.Second, 30*time.Second, func() (bool, error) {
+		podLogOpts := corev1.PodLogOptions{Container: container}
+
+		req := clients.KubeClient.Kube.CoreV1().Pods(ns).GetLogs(podName, &podLogOpts)
+		podLogs, err := req.Stream()
+		if err != nil {
+			return false, err
+		}
+		defer podLogs.Close()
+
+		buf := new(bytes.Buffer)
+		_, err = io.Copy(buf, podLogs)
```

Comment:
```suggestion
		var buf bytes.Buffer
		_, err = io.Copy(&buf, podLogs)
```
Also how much logs does this copy? Until the pod terminates?
We probably want some termination condition?

Reply:
We don't set `Follow` to true in [PodLogOptions](https://godoc.org/k8s.io/api/core/v1#PodLogOptions) so it will just copy the existing logs when the request is made.